pkg/operatorstatus/builder.go

```diff
@@ -0,0 +1,190 @@
+	olm "github.com/operator-framework/operator-lifecycle-manager/pkg/api/apis/operators/v1alpha1"
+	v1 "github.com/operator-framework/operator-marketplace/pkg/apis/operators/v1"
+	v2 "github.com/operator-framework/operator-marketplace/pkg/apis/operators/v2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// NewBuilder returns a builder for ClusterOperatorStatus.
+func NewBuilder(clock clock.Clock) *Builder {
+	return &Builder{
+		clock: clock,
+	}
+}
+
+// Builder helps build ClusterOperatorStatus with appropriate
+// ClusterOperatorStatusCondition and OperandVersion.
+type Builder struct {
+	clock  clock.Clock
+	status *configv1.ClusterOperatorStatus
+}
+
+// GetStatus returns the ClusterOperatorStatus built.
+func (b *Builder) GetStatus() *configv1.ClusterOperatorStatus {
+	return b.status
+}
+
+// WithProgressing sets an OperatorProgressing type condition.
+func (b *Builder) WithProgressing(status configv1.ConditionStatus, message string) *Builder {
+	b.init()
+	condition := &configv1.ClusterOperatorStatusCondition{
+		Type:               configv1.OperatorProgressing,
+		Status:             status,
+		Message:            message,
+		LastTransitionTime: metav1.NewTime(b.clock.Now()),
+	}
+
+	b.setCondition(condition)
+
+	return b
+}
+
+// WithDegraded sets an OperatorDegraded type condition.
+func (b *Builder) WithDegraded(status configv1.ConditionStatus, reason string) *Builder {
+	b.init()
+	condition := &configv1.ClusterOperatorStatusCondition{
+		Type:               configv1.OperatorDegraded,
+		Status:             status,
+		LastTransitionTime: metav1.NewTime(b.clock.Now()),
+		Reason:             reason,
+	}
+
+	b.setCondition(condition)
+
+	return b
+}
+
+// WithAvailable sets an OperatorAvailable type condition.
+func (b *Builder) WithAvailable(status configv1.ConditionStatus, message string) *Builder {
+	b.init()
+	condition := &configv1.ClusterOperatorStatusCondition{
+		Type:               configv1.OperatorAvailable,
+		Status:             status,
+		Message:            message,
+		LastTransitionTime: metav1.NewTime(b.clock.Now()),
+	}
+
+	b.setCondition(condition)
+
+	return b
+}
+
+// WithVersion adds the specific version into the status.
+func (b *Builder) WithVersion(name, version string) *Builder {
+	b.init()
+
+	existing := b.findVersion(name)
+	if existing != nil {
+		existing.Version = version
+		return b
+	}
+
+	ov := configv1.OperandVersion{
+		Name:    name,
+		Version: version,
+	}
+	b.status.Versions = append(b.status.Versions, ov)
+
+	return b
+}
+
+// WithRelatedObject adds the reference specified to the RelatedObjects list.
+func (b *Builder) WithRelatedObject(group, resource, namespace, name string) *Builder {
+	b.init()
+
+	reference := configv1.ObjectReference{
+		Group:     group,
+		Resource:  resource,
+		Namespace: namespace,
+		Name:      name,
+	}
+
+	b.setRelatedObject(reference)
+
+	return b
+}
+
+func (b *Builder) init() {
```

Comment:
This is sort of an anti-pattern when you are writing a builder. I suggest rethinking this.